apm-sniffer/apm-sdk-plugin/websphere-liberty-23.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/websphereliberty/v23/CompleteRunnableInterceptor.java

```diff
@@ -0,0 +1,31 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package org.apache.skywalking.apm.plugin.websphereliberty.v23;
+
+import org.apache.skywalking.apm.agent.core.plugin.interceptor.enhance.EnhancedInstance;
+import org.apache.skywalking.apm.agent.core.plugin.interceptor.enhance.InstanceConstructorInterceptor;
+import org.apache.skywalking.apm.plugin.websphereliberty.v23.async.AsyncType;
+
+public class CompleteRunnableInterceptor implements InstanceConstructorInterceptor {
+    @Override
+    public void onConstruct(final EnhancedInstance objInst, final Object[] allArguments) throws Throwable {
+        // add special flag for AsyncContext distinguish
+        objInst.setSkyWalkingDynamicField(AsyncType.COMPLETE);
```

Comment:
This is a class level attribute. Is this class always staying in asyc completed? No race condition?

Reply:
yes, the `CompleteRunnable` is created in `AsyncContextImpl.complete()` method with AsyncType.COMPLETE attribute and then it will be submit to an executor by `AsyncContextImpl.startWithExecutorThread()`. There are no race condition in the whole process, set and get this attribute is always in one thread.

Reply:
IBM open source liberty code but not release a maven dependency(I can not find it on maven). `CompleteRunnable` and `DispatchRunnable` is not reachable in agent plugin code. That is why I use this strange way rather than `xxx instanceof CompleteRunnable`.

Reply:
You could write classes(fake claases) in weblogic namespace and exclude them in the package stage, then, your codes could work without instanceof and have higher perf.

Reply:
That sounds interesting, I will have a try.

Reply:
Notice the license of those fake files. If you only fake some simple methods, it is fine, but ve clear that they are fake classes and not distributed in their class comments.

Reply:
OK, I will try this.